gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java

```diff
@@ -680,6 +706,76 @@ private boolean ancestorsExistIn(List<? extends TreeItem<Object>> treeItems) {
         return found;
     }
 
+    private MenuItem createCopyProjectNodeItem(List<? extends TreeItem<Object>> selectedTreeItems) {
+        MenuItem copyMenuItem = GseMenuItem.createCopyMenuItem();
+        List<TreeItem<Object>> selectedItems = new ArrayList<>(selectedTreeItems);
+        copyMenuItem.setDisable(ancestorsExistIn(selectedItems) || selectedItems.contains(treeView.getRoot()));
+        copyMenuItem.setOnAction(event -> copy(selectedTreeItems));
+        return copyMenuItem;
+    }
+
+    private MenuItem createPasteProjectNodeItem(TreeItem<Object> selectedTreeItem) {
+        MenuItem pasteMenuItem = GseMenuItem.createPasteMenuItem();
+        final Clipboard clipboard = Clipboard.getSystemClipboard();
+        pasteMenuItem.disableProperty().bind(copied.not());
+        pasteMenuItem.setOnAction(event -> {
+            if (selectedTreeItem.getValue() instanceof ProjectFolder) {
+                ProjectFolder projectFolder = (ProjectFolder) selectedTreeItem.getValue();
+                if (clipboard.hasString()) {
+                    String copyInfos = clipboard.getString();
+                    String[] copyNodesIdArray = copyInfos.split(CopyServiceConstants.PATH_SEPARATOR);
+
+                    // the first index {0} represents the copy signature, the second one represents the copy type
+                    String[] nodeIdArray = ArrayUtils.removeAll(copyNodesIdArray, 0, 1, 2);
+                    Arrays.stream(nodeIdArray).forEach(nodeId ->
+                            copyService.ifPresent(cpService -> {
+                                try {
+                                    cpService.paste(copyNodesIdArray[2], nodeId, projectFolder);
+                                } catch (CopyFailedException ex1) {
+                                    //
+
+                                } catch (CopyNotFinishedException ex2) {
+                                    //
+
+                                } catch (CopyPasteException ex) {
+
+                                }
+                            }));
+                    refresh(selectedTreeItem);
+                }
+            }
+            event.consume();
+        });
+        return pasteMenuItem;
+    }
+
+    private static boolean clipboardContainsProjectNodes(Clipboard systemClipboard) {
+        return systemClipboard.getString().contains(CopyServiceConstants.PROJECTFILE_TYPE) || systemClipboard.getString().contains(CopyServiceConstants.PROJECTFOLDER_TYPE);
+    }
+
+    private void copy(List<? extends TreeItem<Object>> selectedTreeItems) {
+        StringBuilder copyParameters = new StringBuilder();
+
+        List<ProjectNode> projectNodes = selectedTreeItems.stream()
+                .map(item -> (ProjectNode) item.getValue())
+                .collect(Collectors.toList());
+        copyService.ifPresent(cpService -> cpService.copy(projectNodes));
```

Comment:
The creation of copyparameters and the update of the clipboard content should be done after cpService.copy has successfully ended.
Also, the check of the presence of copyService should be done first and throw an exception to the user if not found.

Reply:
Clipboard content should remain empty if the copy failed or did not happen